File: janitor/reprocess_logs.py
#!/usr/bin/python
# Copyright (C) 2018 Jelmer Vernooij <jelmer@jelmer.uk>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

import logging
from datetime import timedelta
from typing import Optional

import silver_platter  # noqa: E402, F401
from buildlog_consultant.common import \
    find_build_failure_description  # noqa: E402
from buildlog_consultant.sbuild import \
    worker_failure_from_sbuild_log  # noqa: E402

from janitor.schedule import do_schedule


def process_sbuild_log(logf):
    failure = worker_failure_from_sbuild_log(logf)
    if failure.error:
        if failure.stage and not failure.error.is_global:
            new_code = f"{failure.stage}-{failure.error.kind}"
        else:
            new_code = failure.error.kind
        try:
            new_failure_details = failure.error.json()
        except NotImplementedError:
            new_failure_details = None
    elif failure.stage:
        new_code = "build-failed-stage-%s" % failure.stage
        new_failure_details = None
    else:
        new_code = "build-failed"
        new_failure_details = None
    new_description = failure.description
    new_phase = failure.phase
    return (new_code, new_description, new_phase, new_failure_details)


def process_build_log(logf):
    lines = [line.decode('utf-8', 'replace') for line in logf]
    match, problem = find_build_failure_description(lines)
    if problem:
        new_code = problem.kind
        try:
            new_failure_details = problem.json()
        except NotImplementedError:
            new_failure_details = None
    else:
        new_code = "build-failed"
        new_failure_details = None
    if match:
        new_description = str(match.line)
    elif problem:
        new_description = str(problem)
    else:
        new_description = "Build failed"
    return (new_code, new_description, "build", new_failure_details)


def process_dist_log(logf):
    lines = [line.decode('utf-8', 'replace') for line in logf]
    problem = find_build_failure_description(lines)[1]
    if problem is None:
        new_code = 'dist-command-failed'
        new_description = "Dist command failed"
        new_failure_details = None
    else:
        if problem.is_global:
            new_code = problem.kind
        else:
            new_code = 'dist-' + problem.kind
        new_description = str(problem)
        try:
            new_failure_details = problem.json()
        except NotImplementedError:
            new_failure_details = None
    new_phase = None
    return (new_code, new_description, new_phase, new_failure_details)


async def reprocess_run_logs(
        db, logfile_manager, *, codebase: str,
        campaign: str, log_id: str, command: str,
        change_set: Optional[str], duration: timedelta,
        result_code: str, description: str, failure_details,
        process_fns, dry_run: bool = False,
        reschedule: bool = False, log_timeout: Optional[int] = None):
    """Reprocess run logs.
    """
    if result_code in ('dist-no-tarball', ):
        return
    for prefix, logname, fn in process_fns:
        if not result_code.startswith(prefix):
            continue
        try:
            logf = await logfile_manager.get_log(
                codebase, log_id, logname, timeout=log_timeout
            )
        except FileNotFoundError:
            return
        else:
            (new_code, new_description, new_phase,
             new_failure_details) = fn(logf)
            break
    else:
        return

    if new_code != result_code or description != new_description or failure_details != new_failure_details:
        logging.info(
            "%s/%s: Updated %r, %r ⇒ %r, %r %r",
            codebase,
            log_id,
            result_code,
            description,
            new_code,
            new_description,
            new_phase
        )
        if not dry_run:
            async with db.acquire() as conn:
                await conn.execute(
                    "UPDATE run SET result_code = $1, description = $2, failure_details = $3 WHERE id = $4",
                    new_code,
                    new_description,
                    new_failure_details,
                    log_id,
                )
                if reschedule and new_code != result_code:
                    await do_schedule(
                        conn,
                        campaign=campaign,
                        change_set=change_set,
                        codebase=codebase,
                        estimated_duration=duration,
                        requestor="reprocess-build-results",
                        bucket="reschedule",
                    )
        return (new_code, new_description, new_failure_details)
    return

File: src/reprocess_logs.rs
use crate::analyze_log::AnalyzeLogFn;
use crate::logs::LogFileManager;
use sqlx::Executor;
use tokio::time::timeout;
use tracing::{error, info};

/// Reprocess run logs.
pub async fn reprocess_run_logs(
    db: &sqlx::PgPool,
    logfile_manager: &dyn LogFileManager,
    codebase: &str,
    campaign: &str,
    log_id: &str,
    _command: &str,
    change_set: Option<&str>,
    duration: chrono::Duration,
    result_code: &str,
    description: &str,
    failure_details: &serde_json::Value,
    process_fns: &[(String, String, AnalyzeLogFn<Box<dyn std::io::Read>>)],
    dry_run: bool,
    reschedule: bool,
    log_timeout: Option<chrono::Duration>,
) -> Option<crate::analyze_log::AnalyzedLog> {
    if ["dist-no-tarball"].contains(&result_code) {
        return None;
    }
    let log_timeout = log_timeout.unwrap_or_else(|| chrono::Duration::minutes(5));
    let mut process_fn_iter = process_fns.iter();
    let new_analysis = loop {
        let (prefix, logname, f) = process_fn_iter.next()?;
        if !result_code.starts_with(prefix) {
            continue;
        }
        match timeout(
            log_timeout.to_std().unwrap(),
            logfile_manager.get_log(codebase, log_id, logname),
        )
        .await
        {
            Ok(r) => match r {
                Ok(logf) => {
                    break f(logf);
                }
                Err(crate::logs::Error::NotFound) => {
                    return None;
                }
                Err(e) => {
                    error!(
                        "{}/{}: Failed to fetch log {}: {:?}",
                        codebase, log_id, logname, e
                    );
                    return None;
                }
            },
            Err(_) => {
                error!("{}/{}: Timeout fetching log {}", codebase, log_id, logname);
                return None;
            }
        }
    };

    if new_analysis.code != result_code
        || new_analysis.description != description
        || new_analysis.failure_details.as_ref() != Some(failure_details)
    {
        info!(
            "{}/{}: Updated {:?}, {:?} ⇒ {:?}, {:?} {:?}",
            codebase,
            log_id,
            result_code,
            description,
            new_analysis.code,
            new_analysis.description,
            new_analysis.phase,
        );
        if !dry_run {
            let query = sqlx::query(
                "UPDATE run SET result_code = $1, description = $2, failure_details = $3 WHERE id = $4")
                .bind(&new_analysis.code)
                .bind(&new_analysis.description)
                .bind(&new_analysis.failure_details)
                .bind(log_id);

            db.execute(query).await.unwrap();
            if reschedule && new_analysis.code != result_code {
                crate::schedule::do_schedule(
                    db,
                    campaign,
                    codebase,
                    "reschedule",
                    change_set,
                    None,
                    false,
                    Some("reprocess-build-results"),
                    Some(duration),
                    None,
                )
                .await
                .unwrap();
            }
        }
        return Some(new_analysis);
    }

    None
}
